smartsim/_core/mli/workermanager.py

```diff
@@ -0,0 +1,716 @@
+        results = [model(tensor) for tensor in transform_result.transformed]
+
+        execute_result = ExecuteResult(results)
+        return execute_result
+
+    @staticmethod
+    def transform_output(
+        request: InferenceRequest,
+        execute_result: ExecuteResult,
+    ) -> OutputTransformResult:
+        transformed = [item.clone() for item in execute_result.predictions]
+        return OutputTransformResult(transformed)
+
+    @staticmethod
+    def serialize_reply(reply: InferenceReply) -> bytes:
+        return pickle.dumps(reply)
+
+
+class ServiceHost(ABC):
+    """Base contract for standalone entrypoint scripts. Defines API for entrypoint
+    behaviors (event loop, automatic shutdown, cooldown) as well as simple
+    hooks for status changes"""
+
+    def __init__(self, as_service: bool = False, cooldown: int = 0) -> None:
+        """Initialize the ServiceHost
+        :param as_service: Determines if the host will run until shutdown criteria
+        are met or as a run-once instance
+        :param cooldown: Period of time to allow service to run before automatic
+        shutdown, in seconds. A non-zero, positive integer."""
+        self._as_service = as_service
+        """If the service should run until shutdown function returns True"""
+        self._cooldown = cooldown
+        """Duration of a cooldown period between requests to the service
+        before shutdown"""
+
+    @abstractmethod
+    def _on_iteration(self, timestamp: int) -> None:
+        """The user-defined event handler. Executed repeatedly until shutdown
+        conditions are satisfied and cooldown is elapsed.
+        :param timestamp: the timestamp at the start of the event loop iteration
+        """
+
+    @abstractmethod
+    def _can_shutdown(self) -> bool:
+        """Return true when the criteria to shut down the service are met."""
+
+    def _on_start(self) -> None:
+        """Empty hook method for use by subclasses. Called on initial entry into
+        ServiceHost `execute` event loop before `_on_iteration` is invoked."""
+        logger.debug(f"Starting {self.__class__.__name__}")
+
+    def _on_shutdown(self) -> None:
+        """Empty hook method for use by subclasses. Called immediately after exiting
+        the main event loop during automatic shutdown."""
+        logger.debug(f"Shutting down {self.__class__.__name__}")
+
+    def _on_cooldown(self) -> None:
+        """Empty hook method for use by subclasses. Called on every event loop
+        iteration immediately upon exceeding the cooldown period"""
+        logger.debug(f"Cooldown exceeded by {self.__class__.__name__}")
+
+    def execute(self) -> None:
+        """The main event loop of a service host. Evaluates shutdown criteria and
+        combines with a cooldown period to allow automatic service termination.
+        Responsible for executing calls to subclass implementation of `_on_iteration`"""
+        self._on_start()
+
+        start_ts = time.time_ns()
```

Comment:
Do we really need ns-granularity?

Reply:
No! I rewrote the automatic shutdown code to simplify it and rely on `datetime.timedelta` instead of getting in the weeds